### PR TITLE
server: check server capacity when start/deploy a vm

### PR DESCRIPTION
**Pull request by weizhouapache**
### Description

This PR fixes the issue that server capacity check is skipped, when start or migrate a vm with specified hostid.

=======UPDATE 2021-09-01===========
We have decided to only add capacity check. the other checks (host state, affinity group, etc) are not added.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [x] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@andrijapanicsb @nvazquez 

the following checks are added when deploy/start a vm with specified id.
(1) host state (Up and Enabled)
(2) DPDK enabled (if it is enabled on vm, host should be enabled also)
(3) hypervisor capability (max guest limits)
(4) server capacity

not sure if users are happy with the checks.

ps: (1) to (3) already exist in destination host check in vm migration. (4) is a new check in vm migration.

**Comment by weizhouapache** [comment]
@blueorangutan test

**Review by nvazquez — COMMENTED** [review]
Thanks @weizhouapache looks good. What about host tags checks against the offering? cc. @andrijapanicsb

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 1]
```diff
@@ -5988,6 +5996,18 @@ public VirtualMachine migrateVirtualMachine(Long vmId, Host destinationHost) thr
...
     }
 
+    private void checkVmDestinationServerCapacity(VMInstanceVO vm, Host destinationHost) {
+        // check server capacity
+        ServiceOfferingVO offering = _offeringDao.findById(vm.getId(), vm.getServiceOfferingId());
+        boolean hostHasCapacity = _capacityMgr.checkIfHostHasCpuCapability(destinationHost.getId(), offering.getCpu(), offering.getSpeed())
+                && _capacityMgr.checkIfHostHasCapacity(destinationHost.getId(), offering.getCpu() * offering.getSpeed(), offering.getRamSize() * 1024L * 1024L, false,
+                _capacityMgr.getClusterOverProvisioningFactor(destinationHost.getClusterId(), Capacity.CAPACITY_TYPE_CPU),
+                _capacityMgr.getClusterOverProvisioningFactor(destinationHost.getClusterId(), Capacity.CAPACITY_TYPE_MEMORY), false);
+        if (! hostHasCapacity) {
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:6006]:
```suggestion
        if (!hostHasCapacity) {
```

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 2]
```diff
@@ -4935,9 +4935,17 @@ public void finalizeStop(VirtualMachineProfile profile, Answer answer) {
...
-            if (!AllowDeployVmIfGivenHostFails.value()) {
-                deployOnGivenHost = true;
+            try {
+                checkVmDestinationServerCapacity(vm, destinationHost);
+                plan = new DataCenterDeployment(vm.getDataCenterId(), destinationHost.getPodId(), destinationHost.getClusterId(), destinationHost.getId(), null, null);
+                if (!AllowDeployVmIfGivenHostFails.value()) {
+                    deployOnGivenHost = true;
+                }
+            } catch (InvalidParameterValueException ex) {
+                s_logger.error("Cannot deploy the VM to specified host due to :" + ex.getMessage());
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:4945]:
```suggestion
                s_logger.error("Cannot deploy the VM to specified host due to: " + ex.getMessage());
```

**Review by sureshanaparti — APPROVED** [review]
clgtm, minor changes to code format required (check suggestions)

**Comment by weizhouapache** [comment]
@nvazquez 
actually vm cannot be started or migrated if there is no enough server capacity in 4.15 and master.

the process to deploy/start a vm without destination host is
(1) check server capacity of all hosts
(2) select a host
(3) double check if host has enough capacity (because step 1 might take long time if there are many hosts). this step was added by #3728

if destination host is specified,  step 1 and 2 are skipped for now. but there is still a new check in step 3.
I am not sure if necessary to add an extra step to check server capacity (it takes few milliseconds) in this PR.

current error 
1. allow.deploy.vm.if.deploy.on.given.host.fails=false
Screenshot from 2021-08-23 10-05-23

2. allow.deploy.vm.if.deploy.on.given.host.fails=true
Screenshot from 2021-08-23 10-09-46

**Comment by weizhouapache** [comment]
> clgtm, minor changes to code format required (check suggestions)

thanks @sureshanaparti , changes have been committed.

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 3]
```diff
@@ -5988,6 +5996,18 @@ public VirtualMachine migrateVirtualMachine(Long vmId, Host destinationHost) thr
         return findMigratedVm(vm.getId(), vm.getType());
     }
 
+    private void checkVmDestinationServerCapacity(VMInstanceVO vm, Host destinationHost) {
+        // check server capacity
+        ServiceOfferingVO offering = _offeringDao.findById(vm.getId(), vm.getServiceOfferingId());
+        boolean hostHasCapacity = _capacityMgr.checkIfHostHasCpuCapability(destinationHost.getId(), offering.getCpu(), offering.getSpeed())
```
Comment (davidjumani) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:6002]: Would it be simpler to use `checkIfHostHasCpuCapabilityAndCapacity`

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 4]
```diff
@@ -4935,9 +4935,17 @@ public void finalizeStop(VirtualMachineProfile profile, Answer answer) {
         boolean deployOnGivenHost = false;
         if (destinationHost != null) {
             s_logger.debug("Destination Host to deploy the VM is specified, specifying a deployment plan to deploy the VM");
-            plan = new DataCenterDeployment(vm.getDataCenterId(), destinationHost.getPodId(), destinationHost.getClusterId(), destinationHost.getId(), null, null);
-            if (!AllowDeployVmIfGivenHostFails.value()) {
-                deployOnGivenHost = true;
+            try {
+                checkVmDestinationServerCapacity(vm, destinationHost);
```
Comment (davidjumani) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:4939]: Would it be simpler to use `_capacityMgr.checkIfHostHasCpuCapabilityAndCapacity` ?

Comment (weizhouapache) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:4939]: @davidjumani  I will change it.thanks

**Comment by weizhouapache** [comment]
@nvazquez @davidjumani updated this pr.

allow.deploy.vm.if.deploy.on.given.host.fails=false
Screenshot from 2021-08-24 10-04-26


allow.deploy.vm.if.deploy.on.given.host.fails=true
Screenshot from 2021-08-24 10-05-16

**Review by davidjumani — APPROVED** [review]

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@blueorangutan test ubuntu18 kvm-centos7

**Comment by nvazquez** [comment]
@blueorangutan test ubuntu18 kvm-ubuntu18

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
Lgtm, failures not related to this PR.
@weizhouapache @nvazquez @borisstoyanov @vladimirpetrov anybody testing this?

**Comment by nvazquez** [comment]
Kicking a new round of tests while is being tested
@blueorangutan test

**Review by vladimirpetrov — APPROVED** [review]
LGTM based on manual testing. Tested scenarios:

(1) cpu speed must be smaller than or equals to cpu mhz of physical host (this does not consider over provisioning)
(2) VM with more cores than available on the host and enough CPU speed and memory
(3) host must have enough cpu resource (vm cpu*speed)
(4) host must have enough memory

**Review by nvazquez — APPROVED** [review]

**Pull request merged**